### PR TITLE
Show error message only if there was an error

### LINKED ISSUE #852: key_app sample application suggests there is an error, even if no error
when using the sample application `key_app`, even on success, it will print:
~~~
!  Last error was: 
~~~

I suggest this line will be printed only if there was an error

### PR DESCRIPTION
**Pull request by RonEld**
This resolves #852. Print the error message, in `key_app`
only if there was an error.

### CODE REVIEW

File: programs/pkey/key_app.c [hunk 1]
```diff
@@ -265,8 +265,11 @@ int main( int argc, char *argv[] )
 exit:
 
 #if defined(MBEDTLS_ERROR_C)
-    mbedtls_strerror( ret, buf, sizeof(buf) );
-    mbedtls_printf( "  !  Last error was: %s\n", buf );
+    if ( ret != 0 )
```
Comment (sbutcher-arm) [inline programs/pkey/key_app.c]: Just before this check to `ret` there are multiple calls to `mbedtls_mpi_write_file`, none of which check their return code. Can you please expand the error checking to the other library calls?

Comment (RonEld) [inline programs/pkey/key_app.c]: Hi @sbutcher-arm I fixed according to your comment

File: programs/pkey/key_app.c [hunk 2]
```diff
@@ -239,34 +239,37 @@ int main( int argc, char *argv[] )
...
 
-exit:
+cleanup:
 
 #if defined(MBEDTLS_ERROR_C)
-    mbedtls_strerror( ret, buf, sizeof(buf) );
-    mbedtls_printf( "  !  Last error was: %s\n", buf );
+    if ( ret != 0 )
+    {
+        mbedtls_strerror( ret, buf, sizeof(buf) );
```
Comment (andresag01) [inline programs/pkey/key_app.c]: Style: Please add space before and after `)` and `(`.

**Comment by sbutcher-arm** [comment]
Somehow this has conflicts. @RonEld - Could you please rebase? Thanks.

**Comment by RonEld** [comment]
The conflicts arose from commit 54ebf9971d1645b2f08623df4b9e40204c4ffd61

@sbutcher-arm I have rebased and resolved conflict
@andresag01 I have fixed the preexisting style issue

**Comment by sbutcher-arm** [comment]
I think we should backport this, as the behaviour in the maintenance branches is simply wrong. @roneld - can you either provide backports or give some justification of why we shouldn't?

cc: @Patater

**Comment by RonEld** [comment]
@sbutcher-arm I don't mind backporting this. We usually don't backport sample application, especially that this bug is just an empty error message, when there is no error.

**Comment by RonEld** [comment]
Backports available at #1799 and #1800 

I have rebased and fixed compilation errors in CI due to new code introduced.
@sbutcher-arm @andresag01 please review

**Comment by sbutcher-arm** [comment]
Note, this looks like it'll conflict with #532, which as a larger PR, I'd like to merge first.

**Comment by RonEld** [comment]
@andresag01 I am not sure I am following you.
I haven't removed any code. Only added the `MBEDTLS_MPI_CHK`.
`mbedtls_export()` and `mbedtls_export_crt()` is still called, but sooner.
The changes you are referring to were introduced in https://github.com/ARMmbed/mbedtls/commit/54ebf9971d1645b2f08623df4b9e40204c4ffd61 , ao I think @hanno-arm can give a better answer

**Comment by andresag01** [comment]
@RonEld: According to the diff, commit 6a9257bc5719dcdcc44dd2c0f52208012b1bffe9 from this PR makes the following kind of changes:
Original:
```
mbedtls_mpi_write_file( "N:  ",  &N,  16, NULL );
```
New:
```
MBEDTLS_MPI_CHK( mbedtls_mpi_write_file( "N:  ", &rsa->N, 16, NULL ) );
```
Note that in the `New` version, we no longer have `&N` but `&rsa->N`, which means that the example is manually inspecting the context of the RSA key instead of using the output of the export functions, which I believe its effectively redundant with the change I pointer out above.

**Comment by RonEld** [comment]
Oh, I see what you mean, Thanks!
It was a merge conflict, I'll fix it ASAP

**Comment by RonEld** [comment]
@andresag01 I have fixed this, please review

**Comment by RonEld** [comment]
@andresag01 Thanks for your approval. You haven't spotted this so late, as it was introduced only today, with the rebase, and it was a difficult one to spot!

**Comment by sbutcher-arm** [comment]
The CI is failing because the FreeBSD build is complaining it can't find `scripts/config.pl`, which is an odd failure. If it went missing I'd expect lots of other, earlier failures, therefore re-running the CI.

**Comment by sbutcher-arm** [comment]
retest

**Comment by sbutcher-arm** [comment]
Needs a Changelog entry, but that can be done at Gatekeeping.

**Pull request merged**